// hypertable/lib/hypertable/index.js
var util = require('util');

exports = module.exports = require('./Client_types');

var Hql_types = require('./Hql_types');
exports.HqlResult = Hql_types.HqlResult;
exports.HqlResult2 = Hql_types.HqlResult2
exports.HqlResultAsArrays = Hql_types.HqlResultAsArrays;

exports.thriftClient = require('./thriftClient');

exports.Key.prototype.toString = function() {
  return util.format('Key(row=\'%s\', column_family=\'%s\', column_qualifier=\'%s\', timestamp=%d, revision=%d, flag=%d)',
                     this.row, 
                     this.column_family,
                     this.column_qualifier,
                     this.timestamp.toNumber(true),
                     this.revision.toNumber(true),
                     this.flag);
};

exports.Cell.prototype.toString = function() {
  return util.format("Cell(key=%s, value='%s')", this.key.toString(), this.value);
};

exports.Timestamp = require('./Timestamp');

exports.SerializedCellsFlag = require('./SerializedCellsFlag.js');
exports.SerializedCellsVersion = require('./SerializedCellsVersion.js');
exports.SerializedCellsWriter = require('./SerializedCellsWriter.js');
