// hypertable/lib/hypertable/index.js
var util = require('util');

exports = module.exports = require('./Client_types');

var Hql_types = require('./Hql_types');
exports.HqlResult = Hql_types.HqlResult;
exports.HqlResult2 = Hql_types.HqlResult2
exports.HqlResultAsArrays = Hql_types.HqlResultAsArrays;
exports.Int64 = require('node-int64');
exports.Timestamp = require('./Timestamp');
exports.SerializedCellsFlag = require('./SerializedCellsFlag.js');
exports.SerializedCellsVersion = require('./SerializedCellsVersion.js');
exports.SerializedCellsReader = require('./SerializedCellsReader.js');
exports.SerializedCellsWriter = require('./SerializedCellsWriter.js');

exports.thriftClient = require('./thriftClient');

exports.Key.prototype.toString = function() {
  var tmpTimestamp;
  var timestampField = '';
  if (this.timestamp !== null && this.timestamp !== undefined) {
    tmpTimestamp = new exports.Timestamp(this.timestamp);
    timestampField = ', timestamp=' + tmpTimestamp.toString();
  }
  var revisionField = '';
  if (this.revision !== null && this.revision !== undefined) {
    tmpTimestamp = new exports.Timestamp(this.revision);
    revisionField = ', revision=' + tmpTimestamp.toString();
  }
  return util.format('Key(row=\'%s\', column_family=\'%s\', column_qualifier=\'%s\'%s%s, flag=%d)',
                     this.row, 
                     this.column_family,
                     this.column_qualifier,
                     timestampField,
                     revisionField,
                     this.flag);
};

exports.Cell.prototype.toString = function() {
  return util.format("Cell(key=%s, value='%s')", this.key.toString(), this.value);
};
